lib/attributor/hash_dsl_compiler.rb

```diff
@@ -0,0 +1,141 @@
+require_relative 'dsl_compiler'
+
+
+module Attributor
+
+  class HashDSLCompiler < DSLCompiler
+
+    # A class that encapsulates the definition of a requirement for Hash attributes
+    # It implements the validation against incoming values and it describes its format for documentation purposes
+    class Requirement
+      attr_reader :type
+      attr_reader :number
+      attr_reader :attr_names
+      attr_reader :description
+
+      def initialize(spec)
+        @description = spec.delete(:description)
+        @type = spec.keys.first
+        case type
+        when :all
+          self.of(*spec[type])
+        when :exclusive
+          self.of(*spec[type])
+        else
+          @number = spec[type]
+        end
+      end
+      def of( *args)
+        @attr_names = args
+        self
+      end
+
+      def validate( object,context=Attributor::DEFAULT_ROOT_CONTEXT,_attribute)
+        result = []
+        case type
+        when :all
+          rest = attr_names - object.keys
+          unless rest.empty?
+            rest.each do |attr|
+              result.push "Key #{attr} is required for #{Attributor.humanize_context(context)}."
+            end
+          end
+        when :exactly
+          included = attr_names & object.keys
+          unless included.size ==  number
+            result.push "Exactly #{number} of the following keys #{attr_names} are required for #{Attributor.humanize_context(context)}. Found #{included.size} instead: #{included.inspect}"
+          end
+        when :at_most
+          rest = attr_names & object.keys
+          if rest.size > number
```

Comment:
same as above.
For example
at_most(2).of a,b,c

for value: a,b,c then
rest=a,b,c therefore rest.size = 3
so, rest.size > 2 = true therefore error